### PR TITLE
Add forms for campaigns

### PR DESCRIPTION
**Pull request by felixfbecker**
Adds a simple form for known campaign types and options.

Note: Going on holiday now so I'll trust @sourcegraph/automation to merge :)

### CODE REVIEW

File: web/src/enterprise/campaigns/detail/forms.tsx [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+const components: Record<CampaignType, React.ComponentType<CampaignFormProps>> = {
+    comby: CombyForm,
+    credentials: CredentialsForm,
+}
+
+export const GenericCampaignForm: React.FunctionComponent<CampaignFormProps & { type: CampaignType }> = ({
+    type,
+    ...props
+}) => {
+    console.log(props.parsed)
```
Comment (mrnugget) [inline web/src/enterprise/campaigns/detail/forms.tsx]:
```suggestion
```

File: web/src/enterprise/campaigns/detail/forms.tsx [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+                    setProperty(campaignArguments, ['matchers', 0], { type: 'npm' }, defaultFormattingOptions)
+                )
+            )
+        } catch {
+            return <ErrorAlert error="Invalid JSON. Please use JSON editor to fix." />
+        }
+        return null
+    }
+    if (parsed.matchers.length > 1) {
+        console.log(parsed)
```
Comment (mrnugget) [inline web/src/enterprise/campaigns/detail/forms.tsx]:
```suggestion
```

File: web/src/enterprise/campaigns/detail/forms.tsx [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+    return (
+        <>
+            {/* <div className="form-group row">
+                <label className="col-sm-2 col-form-label">Type</label>
+                <div className="col-sm-10">
+                    <select className="form-control" value="npm">
+                        <option value="npm">NPM</option>
+                    </select>
+                </div>
+            </div> */}
```
Comment (mrnugget) [inline web/src/enterprise/campaigns/detail/forms.tsx]: Hmmm, is this here on purpose?

File: web/src/enterprise/campaigns/detail/CampaignDetails.tsx [hunk 4]
```diff
@@ -470,25 +471,32 @@ export const CampaignDetails: React.FunctionComponent<Props> = ({
...
+                        <select
+                            className="form-control w-auto d-inline-block e2e-campaign-type"
+                            placeholder="Select campaign type"
+                            onChange={onChangeType}
+                            disabled={!!(campaign && campaign.__typename === 'Campaign')}
+                            value={type}
+                        >
+                            <option value="">Manual</option>
+                            <option value="comby">Comby search and replace</option>
+                            <option value="credentials">NPM Credentials</option>
```
Comment (mrnugget) [inline web/src/enterprise/campaigns/detail/CampaignDetails.tsx]:
```suggestion
                            <option value="credentials">Remove NPM credentials</option>
```

File: web/src/enterprise/campaigns/detail/forms.tsx [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+                <div className="col-sm-10">
+                    <textarea
+                        className="form-control"
+                        value={parsed.matchTemplate}
+                        onChange={createChangeHandler('matchTemplate')}
+                        required={true}
+                    />
+                    <small className="text-muted">
+                        <a rel="noopener noreferrer" target="_blank" href="https://comby.dev/#match-syntax">
+                            Learn about comby syntax
```
Comment (mrnugget) [inline web/src/enterprise/campaigns/detail/forms.tsx]:
```suggestion
                            Learn about Comby's match syntax
```

File: web/src/enterprise/campaigns/detail/forms.tsx [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+                <label className="col-sm-2 col-form-label">Rewrite template</label>
+                <div className="col-sm-10">
+                    <textarea
+                        className="form-control"
+                        value={parsed.rewriteTemplate}
+                        onChange={createChangeHandler('rewriteTemplate')}
+                    />
+                    <small className="text-muted">
+                        <a rel="noopener noreferrer" target="_blank" href="https://comby.dev/#match-syntax">
+                            Learn about comby syntax
```
Comment (mrnugget) [inline web/src/enterprise/campaigns/detail/forms.tsx]: Same suggestion as above. Should we maybe also link to https://catalog.comby.dev/ here?

File: web/src/enterprise/campaigns/detail/forms.tsx (hunk 3 above)
Comment (felixfbecker) [reply web/src/enterprise/campaigns/detail/forms.tsx]: Yes. We only have one type right now. Once we have more, this should be commented in.

File: web/src/enterprise/campaigns/detail/CampaignDetails.tsx (hunk 4 above)
Comment (felixfbecker) [reply web/src/enterprise/campaigns/detail/CampaignDetails.tsx]: This was taken from the current form

File: web/src/enterprise/campaigns/detail/forms.tsx (hunk 5 above)
Comment (felixfbecker) [reply web/src/enterprise/campaigns/detail/forms.tsx]: This is the same as in the current form

File: web/src/enterprise/campaigns/detail/CampaignDetails.tsx (hunk 4 above)
Comment (mrnugget) [reply web/src/enterprise/campaigns/detail/CampaignDetails.tsx]: Sure, I've seen it in the current form. This was just a suggestion to improve the form while you're working on it anyway :)

Comment (felixfbecker) [reply web/src/enterprise/campaigns/detail/CampaignDetails.tsx]: Well I’m on vacation now so I’d prefer to discuss any other improvement as a follow-up once I’m back, if that’s okay 🙂

Comment (mrnugget) [reply web/src/enterprise/campaigns/detail/CampaignDetails.tsx]: Sure! This wasn't meant as a "please fix this before you head off into vacation", but just a response to the request for review that I received.

**Comment by felixfbecker** [comment]
Closing for now

**Pull request closed without merging**